python/lsst/cp/pipe/cpFunctions.py

```diff
@@ -0,0 +1,682 @@
+                    scale = [inputScales[detectorId][amp.getName()][visitId] for amp in exp.getDetector()]
+                else:
+                    raise RuntimeError(f"Unknown scaling level: {self.config.scalingLevel}")
+            else:
+                raise RuntimeError(f"Unknown scaling type: {self.config.exposureScaling}.")
+
+            expScales.append(scale)
+            self.log.info("Scaling input %d by %s", index, scale)
+            self.applyScale(exp, scale)
+
+        self.combine(combined, inputExps, stats)
+
+        if np.isfinite(self.config.outputScaling) and self.config.outputScaling != 0.0:
+            background = self.stats.run(combined)
+            self.log.info("Measured background of stack is %f; adjusting to %f" %
+                          (background, self.config.outputScaling))
+
+            self.applyScale(combinedExp, float(background) / float(self.config.outputScaling))
+            expScales = [scl * self.config.outputScaling / background for scl in expScales]
+
+        self.interpolateNans(combined)
+
+        # Combine headers
+        self.combineHeaders(inputExps, combinedExp,
+                            calibType=self.config.calibrationType, scales=expScales)
+
+        # Return
+        return pipeBase.Struct(
+            outputData=combinedExp,
+        )
+
+    def getDimensions(self, expList):
+        """Get dimensions of the inputs.
+
+        Parameters
+        ----------
+        expList : `list` [`lsst.afw.image.Exposure`]
+            Exps to check the sizes of.
+
+        Returns
+        -------
+        width, height : `int`
+            Unique set of input dimensions.
+        """
+        dimList = [exp.getDimensions() for exp in expList if exp is not None]
+        return self.getSize(dimList)
+
+    def getSize(self, dimList):
+        """Determine a consistent size, given a list of image sizes.
+
+        Parameters
+        -----------
+        dimList : iterable of `tuple` (`int`, `int`)
+            List of dimensions.
+
+        Raises
+        ------
+        RuntimeError
+            If input dimensions are inconsistent.
+
+        Returns
+        --------
+        width, height : `int`
+            Common dimensions.
+        """
+        dim = set((w, h) for w, h in dimList)
+        if len(dim) != 1:
+            raise RuntimeError("Inconsistent dimensions: %s" % dim)
+        return dim.pop()
+
+    def applyScale(self, exposure, scale=None):
+        """Apply scale to input exposure.
+
+        This implementation applies a flux scaling: the input exposure is
+        divided by the provided scale.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to scale.
+        scale : `float` or `list` [`float`], optional
+            Constant scale to divide the exposure by.
+        """
+        if scale is not None:
+            mi = exposure.getMaskedImage()
+            if isinstance(scale, list):
```

Comment:
You have to be a little careful with strings:
```python
>>> import collections
>>> isinstance("xyz", collections.Iterable)
__main__:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
True
```
(noting that it's `collections.abc.Iterable` these days)